### PR TITLE
Add PCMDI Diags to zppy

### LINKED ISSUE #141: Add task for PMP
Add task for PCMDI Metrics Package (PMP): https://climatemodeling.science.energy.gov/e3sm/pmp-0181-0200

### PR DESCRIPTION
**Pull request by forsyth2**
## Summary

Objectives:
- Add PCMDI Diags to zppy
- Follow-up to #656, #718
- Corresponding `zppy-interfaces` PR: https://github.com/E3SM-Project/zppy-interfaces/pull/25

Issue resolution:
- Closes #141

Select one: This pull request is...
- [ ] a bug fix: increment the patch version
- [ ] a small improvement: increment the minor version
- [x] a new feature: increment the minor version
- [ ] an incompatible (non-backwards compatible) API change: increment the major version

## Big Change

- [ ] To merge, I will use "Create a merge commit". That is, this change is large enough to require multiple units of work (i.e., it should be multiple commits).

### 1. Does this do what we want it to do?

Required:
- [ ] Product Management: I have confirmed with the stakeholders that the objectives above are correct and complete.
- [ ] Testing: I have added or modified at least one "min-case" configuration file to test this change. Every objective above is represented in at least one cfg.
- [ ] Testing: I have considered likely and/or severe edge cases and have included them in testing.

If applicable:
- [ ] Testing: this pull request introduces an important feature or bug fix that we _must_ test often. I have updated the weekly-test configuration files, not just a "min-case" one.
- [ ] Testing: this pull request adds at least one new possible parameter to the cfg. I have tested using this parameter with and without any other parameter that may interact with it.

### 2. Are the implementation details accurate & efficient?

Required:
- [ ] Logic: I have visually inspected the entire pull request myself.
- [ ] Logic: I have left GitHub comments highlighting important pieces of code logic. I have had these code blocks reviewed by at least one other team member.

If applicable:
- [ ] Dependencies: This pull request introduces a new dependency. I have discussed this requirement with at least one other team member. The dependency is noted in `zppy/conda`, not just an `import` statement.

### 3. Is this well documented?

Required:
- [ ] Documentation: by looking at the docs, a new user could easily understand the functionality introduced by this pull request.

### 4. Is this code clean?

Required:
- [ ] Readability: The code is as simple as possible and well-commented, such that a new team member could understand what's happening.
- [ ] Pre-commit checks: All the pre-commits checks have passed.

If applicable:
- [ ] Software architecture: I have discussed relevant trade-offs in design decisions with at least one other team member. It is unlikely that this pull request will increase tech debt.

### CODE REVIEW

**Comment by forsyth2** [comment]
The first commit here, aee525ae91890e6f6b96b1f63715ed56ae270390, was created with:
```bash
# 2025-05-22
git fetch zhangshixuan1987 zppy_pcmdi_new
git checkout -b zppy_pcmdi_202505 zhangshixuan1987/zppy_pcmdi_new
git log
# The latest commit not from this PR is "Test fixes (#646)", from 11/25
# So, this branch is now ~6 months out of date. 
# Very large changes to the zppy codebase have occurred over this time.
git rebase -i 69fe79dd718832f52d7a0fae31c0257b9ee70915
# Usally, we like to keep the commit history for big changes,
# but considering we need to catch up on 6 months of code changes,
# it will be much more practical to rebase the 24 commits into 1.
# Then, we only have to rebase 1 commit.
# So, setting "pick" for the 1st commit "Add PCMDI Diags to zppy"
# and "f" for "fixup" for the remaining 23 commits.
git log
# Now, we just have one commit on top of "Test fixes (#646)"
# So now let's address the conflicts from rebasing off the latest main.
git fetch upstream
git rebase upstream/main
# Luckily, only two files have conflicts
# Just did "Accept Incoming Change" on all conflicts.
git grep -n "<<<<<<<"
# No remaining conflicts
git add zppy/defaults/default.ini
git add zppy/templates/ts.bash
git status 
# All changes are ready to be committed
git rebase --continue
# Now, we have one commit on top of "Update to latest conda-incubator/setup-miniconda (#715)", from 5/21.
# So, we're up to date now!
```

**Comment by forsyth2** [comment]
Second commit, f769b9382469a98a2c6996bcfb5bed43779c772e, created by iteratively fixing errors brought up by running `pre-commit run --all-files`. Many of these errors were because of function renamings during parameter inference refactoring (#682)

**Comment by forsyth2** [comment]
3rd commit, dff9099b505d737c85c2fd7c7dfdcc38a0451276, cleans up the file space.

**Comment by forsyth2** [comment]
4th commit, cab9aa23a7611e93130fa790bab4e968e9bf1d84, moves the `ts` task changes to the `e3sm_to_cmip` task.

**Comment by forsyth2** [comment]
5th commit, 0e2ef3d8f50a62a4d659779243958f0fa1c7630a removes unused parameters.

6th commit, a0c9c5f38ba5852ad13659a6088a9a9fb269d77e, puts `pcmdi_diags` task at bottom of `default.ini`, after `ilamb`, since that's the order it appears in `__main__.py`.

**Review by forsyth2 — COMMENTED** [review]
I've done a high-level code read-through, leaving some comments.

I'd also like to:
1. Run a sample `cfg` producing PCMDI plots
2. Run the integration tests to make sure the other tasks haven't been broken by these changes.

**Comment by forsyth2** [comment]
6c6818edd49770307c4b7f62a490cd75b364dc83 adds the example cfg listed in https://github.com/E3SM-Project/zppy/pull/656#issuecomment-2832487957

**Comment by forsyth2** [comment]
5003ca038c20a1e6ae20cfe0ea53a35eb1c3a4eb adds some initial edits to the example cfg. I'm still not able to run it, because the new task requires both the Unified environment for the `nc`-file operators _and_ an environment with the PCMDI code. I think it will be easier to reason about the environments post-refactor (moving some of the code to `zppy-interfaces`), so I'm going to move onto prototyping that work.

File: examples/post.v3.LR.amip.0101.cfg [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+[e3sm_to_cmip]
+active = True
+frequency = "monthly"
+ts_num_years=5
+walltime = "00:10:00"
+years = "2005:2015:5"
+
+  [[ atm_monthly_180x360_aave ]]
+  # list of variables to be interpolated to pressure level; see e3sm_to_cmip/vrt_remap_plev19.nc
+  cmip_plevdata = "inclusions/e3sm_to_cmip/vrt_remap_plev19.nc"
```
Comment (forsyth2) [inline examples/post.v3.LR.amip.0101.cfg:36]: This file doesn't exist in `zppy`'s `zppy/templates/inclusions/e3sm_to_cmip` path. Is it `inclusions` elsewhere?

File: zppy/templates/pcmdi_diags.bash [hunk 2]
```diff
@@ -0,0 +1,852 @@
...
+    local ttag tmp_file MM combined_name
+
+    fname=$(basename "${file}")
+
+    # Match two date patterns (YYYYMM or YYYYMMDD) separated by _ or -
+    if [[ ${fname} =~ ([0-9]{6,8})[_-]([0-9]{6,8}) ]]; then
+      YYYYS="${BASH_REMATCH[1]}"
+      YYYYE="${BASH_REMATCH[2]}"
+    else
+      echo "Warning: Could not extract dates from ${fname}"
```
Comment (forsyth2) [inline zppy/templates/pcmdi_diags.bash:197]: Copy of https://github.com/E3SM-Project/zppy/pull/719/files#r2141246726:
```
> cd /lcrc/group/e3sm/ac.forsyth2/zppy_pr719_output/unique_id_12/v3.LR.amip_0101/post/scripts
> emacs pcmdi_diags_enso_model_vs_obs_2005-2015.o753616
# Warning: Could not extract dates from *.nc
# Warning: Catalogue not found at pcmdi_diags/ts_enso_catalogue.json
# Warning: Catalogue not found at pcmdi_diags/ts_ref_enso_catalogue.json
# KeyError: "Variable key 'psl' not found in observation dictionary."

> emacs pcmdi_diags_mean_climate_model_vs_obs_2005-2014.o753613 
# Warning: Could not extract dates from *.nc
# Warning: Catalogue not found at pcmdi_diags/climo_mean_climate_catalogue.json
# Warning: Catalogue not found at pcmdi_diags/climo_ref_mean_climate_catalogue.json
# ValueError: not enough values to unpack (expected 4, got 1)

> emacs pcmdi_diags_variability_modes_atm_model_vs_obs_2005-2014.o753615 
# Warning: Could not extract dates from *.nc
# Warning: Catalogue not found at pcmdi_diags/ts_variability_modes_atm_catalogue.json
# Warning: Catalogue not found at pcmdi_diags/ts_ref_variability_modes_atm_catalogue.json
# KeyError: 'psl'

> emacs pcmdi_diags_variability_modes_cpl_model_vs_obs_2005-2014.o753614
# Warning: Could not extract dates from *.nc
# Warning: Catalogue not found at pcmdi_diags/ts_variability_modes_cpl_catalogue.json
# Warning: Catalogue not found at pcmdi_diags/ts_ref_variability_modes_cpl_catalogue.json
# KeyError: 'ts'
```

File: zppy/templates/pcmdi_diags.bash [hunk 3]
```diff
@@ -0,0 +1,852 @@
...
+  local file fname PREFIX YYYYS YYYYE ttag combined_name
+
+  for file in ${ts_dir_source}/*.nc; do
+    fname=$(basename "$file")
+    # Match two time patterns (YYYYMM or YYYYMMDD) separated by _ or -
+    if [[ ${fname} =~ ([0-9]{6,8})[_-]([0-9]{6,8}) ]]; then
+      YYYYS="${BASH_REMATCH[1]}"
+      YYYYE="${BASH_REMATCH[2]}"
+    else
+      echo "Warning: Could not extract dates from ${fname}"
```
Comment (forsyth2) [inline zppy/templates/pcmdi_diags.bash:336]: Copy of https://github.com/E3SM-Project/zppy/pull/719/files#r2141246726:
```
> cd /lcrc/group/e3sm/ac.forsyth2/zppy_pr719_output/unique_id_12/v3.LR.amip_0101/post/scripts
> emacs pcmdi_diags_enso_model_vs_obs_2005-2015.o753616
# Warning: Could not extract dates from *.nc
# Warning: Catalogue not found at pcmdi_diags/ts_enso_catalogue.json
# Warning: Catalogue not found at pcmdi_diags/ts_ref_enso_catalogue.json
# KeyError: "Variable key 'psl' not found in observation dictionary."

> emacs pcmdi_diags_mean_climate_model_vs_obs_2005-2014.o753613 
# Warning: Could not extract dates from *.nc
# Warning: Catalogue not found at pcmdi_diags/climo_mean_climate_catalogue.json
# Warning: Catalogue not found at pcmdi_diags/climo_ref_mean_climate_catalogue.json
# ValueError: not enough values to unpack (expected 4, got 1)

> emacs pcmdi_diags_variability_modes_atm_model_vs_obs_2005-2014.o753615 
# Warning: Could not extract dates from *.nc
# Warning: Catalogue not found at pcmdi_diags/ts_variability_modes_atm_catalogue.json
# Warning: Catalogue not found at pcmdi_diags/ts_ref_variability_modes_atm_catalogue.json
# KeyError: 'psl'

> emacs pcmdi_diags_variability_modes_cpl_model_vs_obs_2005-2014.o753614
# Warning: Could not extract dates from *.nc
# Warning: Catalogue not found at pcmdi_diags/ts_variability_modes_cpl_catalogue.json
# Warning: Catalogue not found at pcmdi_diags/ts_ref_variability_modes_cpl_catalogue.json
# KeyError: 'ts'
```

Comment (forsyth2) [reply zppy/templates/pcmdi_diags.bash:336]: `fname=$(basename "$file")` is just getting the regex pattern `*.nc` as a string, so it's not actually matching individual files. To check -- is that accounted for elsewhere in the code or is this a bug?

**Comment by forsyth2** [comment]
Closing in favor of #732. (Note the zppy-interfaces PR (https://github.com/E3SM-Project/zppy-interfaces/pull/25) remains unchanged.
